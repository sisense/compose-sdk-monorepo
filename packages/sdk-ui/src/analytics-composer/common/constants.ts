export const CODE_TEMPLATES_INDENT = 6;
