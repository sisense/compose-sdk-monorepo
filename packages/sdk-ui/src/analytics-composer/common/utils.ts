import { TranslatableError } from '@/translation/translatable-error';
import { ExpandedQueryModel, SimpleQueryModel } from '../types.js';

export function toKebabCase(str: string): string {
  return str
    .replace(/\s+/g, '-') // Replace whitespace with hyphens
    .replace(/([a-z])([A-Z])/g, '$1-$2') // Convert camelCase to kebab-case
    .toLowerCase(); // Convert to lowercase
}

export function capitalizeFirstLetter(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

/**
 * Sanitize ID of a dimension
 *
 * @param str - input string
 * @return sanitized ID
 */
export function sanitizeDimensionId(str: string) {
  // Regular expression pattern to match [table.column (Calendar)]
  const regex = /\[(.+?)\s\(Calendar\)\]/;

  // Replace [table.column (Calendar)] with [table.column]
  return str.replace(regex, '[$1]').trim();
}

export function validateQueryModel(model: any): SimpleQueryModel {
  if (!model) {
    throw new TranslatableError('errors.emptyModel');
  } else if (!model.metadata) {
    throw new TranslatableError('errors.missingMetadata');
  } else if (!model.model) {
    throw new TranslatableError('errors.missingModelTitle');
  }

  return model;
}

export function isEmptyQueryModel(queryModel: ExpandedQueryModel | undefined | null): boolean {
  return !queryModel || !queryModel.jaql.datasource.title || !queryModel.jaql.metadata.length;
}

export function escapeSingleQuotes(str?: string) {
  if (!str) {
    return str;
  }
  // Replace single quotes with escaped single quotes
  // Only when the single quote is not preceded by a backslash
  return str.replace(/(?<!\\)'/g, "\\'");
}
