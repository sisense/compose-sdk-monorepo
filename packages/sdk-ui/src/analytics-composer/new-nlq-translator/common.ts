import {
  createAttributeHelper,
  DateLevels,
  Filter,
  FilterRelations,
  JaqlDataSourceForDto,
  filterFactory,
  measureFactory,
  JSONArray,
  JSONValue,
  Measure,
  DataType,
} from '@sisense/sdk-data';
import { NormalizedTable, NormalizedField } from '../types.js';

// Type to represent a function arg
export type ParsedArg = string | number | ParsedFunctionCall | string[];

// Type to represent the structure parsed from the JSON
export type ParsedFunctionCall = {
  function: string;
  args: ParsedArg[];
};

function isParsedFunctionCall(value: JSONValue): value is ParsedFunctionCall {
  return typeof value === 'object' && value !== null && 'function' in value && 'args' in value;
}

export function isParsedFunctionCallArray(value: JSONArray): value is ParsedFunctionCall[] {
  return value.every(isParsedFunctionCall);
}

type QueryElement = Filter | FilterRelations | Measure;

// Type guard to check if the argument is a Filter
export function isFilterElement(arg: QueryElement): arg is Filter {
  return 'attribute' in arg && 'config' in arg;
}

// Type guard to check if the argument is a FilterRelations
export function isFilterRelationsElement(arg: QueryElement): arg is FilterRelations {
  return 'left' in arg && 'right' in arg && 'operator' in arg;
}

export function isMeasureElement(arg: QueryElement): arg is Measure {
  return !isFilterElement(arg) && !isFilterRelationsElement(arg);
}

export function getFactoryFunctionsMap(
  factoryName: string,
  factoryObject: Record<string, any>,
): Record<string, (...args: any[]) => any> {
  const functionsMap: Record<string, (...args: any[]) => any> = {};
  const prefix = factoryName !== '' ? factoryName + '.' : '';
  for (const key in factoryObject) {
    if (typeof factoryObject[`${key}`] === 'function') {
      functionsMap[`${prefix}${key}`] = factoryObject[`${key}`];
    } else if (typeof factoryObject[`${key}`] === 'object') {
      const nestedFunctionsMap = getFactoryFunctionsMap('', factoryObject[`${key}`]);
      for (const nestedKey in nestedFunctionsMap) {
        functionsMap[`${prefix}${key}.${nestedKey}`] = nestedFunctionsMap[`${nestedKey}`];
      }
    }
  }
  return functionsMap;
}

export const measureFactoryFunctionsMap = getFactoryFunctionsMap('measureFactory', measureFactory);
export const filterFactoryFunctionsMap = getFactoryFunctionsMap('filterFactory', filterFactory);

function executeFunction(functionPath: string, args: any[]): QueryElement {
  const parts = functionPath.split('.');
  const factoryName = parts[0];

  if (factoryName === 'filterFactory' && filterFactoryFunctionsMap[`${functionPath}`]) {
    return (
      filterFactoryFunctionsMap[`${functionPath}`] as (...params: any[]) => Filter | FilterRelations
    )(...args);
  } else if (factoryName === 'measureFactory' && measureFactoryFunctionsMap[`${functionPath}`]) {
    return (measureFactoryFunctionsMap[`${functionPath}`] as (...params: ParsedArg[]) => Measure)(
      ...args,
    );
  }

  throw new Error(`Function "${functionPath}" not found in filterFactory or measureFactory.`);
}

function parseAttributeName(attributeName: string): {
  tableName: string;
  columnName: string;
  level?: string;
} {
  // Parse "DM.Commerce.Date.Years" -> tableName: "Commerce", columnName: "Date", level: "Years"
  // Parse "DM.Brand.Brand" -> tableName: "Brand", columnName: "Brand", level: undefined
  const parts = attributeName.split('.');
  if (parts.length < 3 || parts[0] !== 'DM') {
    throw new Error(
      `Invalid attribute name format: "${attributeName}". Expected format: "DM.TableName.ColumnName[.Level]".`,
    );
  }

  const tableName = parts[1];
  const columnName = parts[2];
  let level: string | undefined;

  // extract level if present (4th part)
  if (parts.length === 4) {
    level = parts[3];
  }

  return { tableName, columnName, level };
}

function findDataFieldAndLevel(
  attributeName: string,
  tables: NormalizedTable[],
): { field: NormalizedField; level?: string } {
  const { tableName, columnName, level } = parseAttributeName(attributeName);

  const table = tables.find((t) => t.name === tableName);
  if (!table) {
    throw new Error(`Table "${tableName}" not found in the data model.`);
  }

  const field = table.columns.find((column) => column.name === columnName);
  if (!field) {
    throw new Error(`Column "${columnName}" not found in table "${tableName}".`);
  }

  if (level) {
    if (field.dataType !== DataType.DATETIME) {
      throw new Error(
        `Invalid date level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is not a datetime column.`,
      );
    }
    const validLevels = DateLevels.all;
    if (!validLevels.includes(level)) {
      throw new Error(
        `Invalid date level "${level}" in attribute "${attributeName}". Valid levels are: ${validLevels.join(
          ', ',
        )}.`,
      );
    }
  }

  return { field, level };
}

export function createAttribute(
  attributeName: string,
  dataSource: JaqlDataSourceForDto,
  tables: NormalizedTable[],
) {
  const { field, level } = findDataFieldAndLevel(attributeName, tables);

  return createAttributeHelper({
    expression: field.expression,
    dataType: field.dataType,
    granularity: level,
    format: undefined,
    sort: undefined,
    dataSource,
  });
}

function processArg(arg: ParsedArg, dataSource: JaqlDataSourceForDto, tables: NormalizedTable[]) {
  if (
    typeof arg === 'object' &&
    arg !== null &&
    !Array.isArray(arg) &&
    'function' in arg &&
    'args' in arg
  ) {
    return processNode(arg, dataSource, tables);
  }
  if (typeof arg === 'string' && arg.startsWith('DM.')) {
    return createAttribute(arg, dataSource, tables);
  }
  return arg;
}

/**
 * Recursively processes a parsed JSON object to execute filterFactory and measureFactory functions.
 *
 * @param parsedObject - The parsed JSON object representing a function call or a nested structure.
 * @param dataSource - The data source to use for the query.
 * @param tables - The tables to use for the query.
 * @returns The result of the executed function or the processed nested structure.
 */
export function processNode(
  parsedObject: ParsedFunctionCall,
  dataSource: JaqlDataSourceForDto,
  tables: NormalizedTable[],
): QueryElement {
  const { function: functionPath, args } = parsedObject;
  // Recursively process arguments that are also function calls
  const processedArgs = args.map((arg) => processArg(arg, dataSource, tables));
  return executeFunction(functionPath, processedArgs);
}
