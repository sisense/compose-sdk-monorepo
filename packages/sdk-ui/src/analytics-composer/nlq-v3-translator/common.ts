import {
  Attribute,
  createAttributeHelper,
  DateLevels,
  isDatetime,
  JaqlDataSourceForDto,
  MetadataTypes,
} from '@sisense/sdk-data';

import { NlqTranslationResult, NormalizedField, NormalizedTable } from '../types.js';
import { ATTRIBUTE_PREFIX } from './types.js';

function parseAttributeName(attributeName: string): {
  tableName: string;
  columnName: string;
  level?: string;
} {
  // Parse "DM.Commerce.Date.Years" -> tableName: "Commerce", columnName: "Date", level: "Years"
  // Parse "DM.Brand.Brand" -> tableName: "Brand", columnName: "Brand", level: undefined
  const parts = attributeName.split('.');
  if (parts.length < 3 || `${parts[0]}.` !== ATTRIBUTE_PREFIX) {
    throw new Error(
      `Invalid attribute name format: "${attributeName}". Expected format: "${ATTRIBUTE_PREFIX}TableName.ColumnName[.Level]"`,
    );
  }

  const tableName = parts[1];
  const columnName = parts[2];
  let level: string | undefined;

  // extract level if present (4th part)
  if (parts.length === 4) {
    level = parts[3];
  }

  return { tableName, columnName, level };
}

function findDataFieldAndLevel(
  attributeName: string,
  tables: NormalizedTable[],
): { field: NormalizedField; level?: string } {
  const { tableName, columnName, level } = parseAttributeName(attributeName);

  const table = tables.find((t) => t.name === tableName);
  if (!table) {
    throw new Error(`Table "${tableName}" not found in the data model`);
  }

  const field = table.columns.find((column) => column.name === columnName);
  if (!field) {
    throw new Error(`Column "${columnName}" not found in table "${tableName}"`);
  }

  if (level) {
    if (!isDatetime(field.dataType)) {
      throw new Error(
        `Invalid date level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is not a datetime column`,
      );
    }
    const validLevels = DateLevels.all;
    if (!validLevels.includes(level)) {
      throw new Error(
        `Invalid date level "${level}" in attribute "${attributeName}". Valid levels are: ${validLevels.join(
          ', ',
        )}`,
      );
    }

    // throw error for time levels on date only column
    const dateOnlyLevels = DateLevels.dateOnly;
    if (field.dataType === 'date' && !dateOnlyLevels.includes(level)) {
      throw new Error(
        `Invalid level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is only a date column, not a datetime column`,
      );
    }

    // throw error for date levels on time only column
    const timeOnlyLevels = DateLevels.timeOnly;
    if (field.dataType === 'time' && !timeOnlyLevels.includes(level)) {
      throw new Error(
        `Invalid level "${level}" in attribute "${attributeName}". Column "${tableName}.${columnName}" is only a time column, not a date column`,
      );
    }
  }

  return { field, level };
}

/**
 * Validates that an attribute is compatible with text values
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is a text type
 */
export function isTextAttribute(attribute: Attribute): boolean {
  return attribute.type === MetadataTypes.TextAttribute;
}

/**
 * Validates that an attribute is compatible with numeric values
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is a numeric type
 */
export function isNumericAttribute(attribute: Attribute): boolean {
  return attribute.type === MetadataTypes.NumericAttribute;
}

/**
 * Validates that an attribute is a date/datetime type
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is a date/datetime type
 */
export function isDateLevelAttribute(attribute: Attribute): boolean {
  return attribute.type === MetadataTypes.DateLevel;
}

/**
 * Validates that an attribute is compatible with text OR numeric values
 * Used for filters like equals, doesntEqual that accept both text and number
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is text or numeric type
 */
export function isTextOrNumericAttribute(attribute: Attribute): boolean {
  return isTextAttribute(attribute) || isNumericAttribute(attribute);
}

/**
 * Validates that an attribute is NOT a date type
 * Used for filters that should not work with date attributes
 *
 * @param attribute - The attribute to validate
 * @returns True if the attribute is not a date type
 */
export function isNonDateLevelAttribute(attribute: Attribute): boolean {
  return !isDateLevelAttribute(attribute);
}

/**
 * Gets a display string for the attribute type
 *
 * @param attribute - The attribute to get the type display string for
 * @returns A human-readable string representing the attribute type
 */
export function getAttributeTypeDisplayString(attribute: Attribute): string {
  return attribute.type === MetadataTypes.TextAttribute
    ? 'text'
    : attribute.type === MetadataTypes.NumericAttribute
    ? 'numeric'
    : attribute.type === MetadataTypes.DateLevel
    ? 'date/datetime'
    : 'unknown';
}

export function createAttributeFromName(
  attributeName: string,
  dataSource: JaqlDataSourceForDto,
  tables: NormalizedTable[],
) {
  const { field, level } = findDataFieldAndLevel(attributeName, tables);

  return createAttributeHelper({
    expression: field.expression,
    dataType: field.dataType,
    granularity: level,
    format: undefined,
    sort: undefined,
    dataSource,
  });
}

export function getSuccessData<T>(result: NlqTranslationResult<T>): T {
  if (!result.success) throw new Error('Expected success result');
  return result.data;
}

export function getErrors<T>(result: NlqTranslationResult<T>): string[] {
  if (result.success) throw new Error('Expected error result');
  return result.errors.map((error) => error.message);
}
