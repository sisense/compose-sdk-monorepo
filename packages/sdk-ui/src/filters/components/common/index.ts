export { LinkButton, PrimaryButton, SecondaryButton } from './Buttons';
export { Checkbox } from './Checkbox';
export { Input } from './Input';
export { Radio } from './Radio';
export { SisenseSwitchButton } from './Switch';
