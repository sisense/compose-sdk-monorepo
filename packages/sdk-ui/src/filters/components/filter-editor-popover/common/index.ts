export * from './input';
export * from './selectable-section';
export * from './search-input';
export * from './select';
