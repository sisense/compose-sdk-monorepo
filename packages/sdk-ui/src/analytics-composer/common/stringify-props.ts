import { DataSource, Filter, FilterRelations } from '@sisense/sdk-data';
import { TranslatableError } from '@/translation/translatable-error';
import { toKebabCase, isNonEmptyArray } from './utils';
import { stringifyProps } from '../widget/stringify-props';
import { ChartDataOptions } from '@/types';

export function stringifyDataOptions(dataOptions: ChartDataOptions): string {
  return stringifyProps(dataOptions, 6);
}

export function stringifyDataSource(dataSource: DataSource | undefined): string {
  if (!dataSource) {
    throw new TranslatableError('errors.undefinedDataSource');
  }

  let dataSourceString: string;

  if (typeof dataSource === 'object' && 'title' in dataSource) {
    dataSourceString = dataSource.title;
  } else {
    dataSourceString = dataSource;
  }

  return toKebabCase(dataSourceString);
}

export function stringifyExtraImports(
  filters: Filter[] | FilterRelations,
  importMeasureFactory = true,
): string {
  const importNames = [];

  if (importMeasureFactory) {
    importNames.push('measureFactory');
  }

  if (isNonEmptyArray(filters as Filter[])) {
    importNames.push('filterFactory');
  }

  return importNames.length > 0
    ? `import { ${importNames.join(', ')} } from '@sisense/sdk-data';`
    : '';
}
