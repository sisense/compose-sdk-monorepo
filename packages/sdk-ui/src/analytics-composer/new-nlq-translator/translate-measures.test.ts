import { describe, it, expect } from 'vitest';
import {
  translateMeasuresJSON,
  translateMeasuresFromJSONFunctionCall,
} from './translate-measures.js';

import {
  MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
  MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
} from '../__mocks__/mock-data-sources.js';
import { ParsedFunctionCall } from './common.js';

describe('translateMeasures', () => {
  describe('translateMeasuresFromJSONFunctionCall', () => {
    it('should translate measures from parsed function calls', () => {
      const mockMeasuresJSON: ParsedFunctionCall[] = [
        { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue', 'Total Revenue'] },
        { function: 'measureFactory.count', args: ['DM.Commerce.Revenue', 'Revenue Count'] },
      ];

      const measures = translateMeasuresFromJSONFunctionCall(
        mockMeasuresJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(measures).toHaveLength(2);
      expect(measures).toMatchSnapshot();
    });

    it('should translate single measure', () => {
      const mockMeasuresJSON: ParsedFunctionCall[] = [
        { function: 'measureFactory.average', args: ['DM.Commerce.Revenue'] },
      ];

      const measures = translateMeasuresFromJSONFunctionCall(
        mockMeasuresJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(measures).toHaveLength(1);
      expect(measures).toMatchSnapshot();
    });

    it('should throw error when trying to use filter function as measure', () => {
      const mockMeasuresJSON: ParsedFunctionCall[] = [
        { function: 'filterFactory.members', args: ['DM.Country.Country', ['United States']] },
      ];

      expect(() => {
        translateMeasuresFromJSONFunctionCall(
          mockMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow('Invalid measure JSON');
    });

    it('should throw error when trying to use filter relation as measure', () => {
      const mockMeasuresJSON: ParsedFunctionCall[] = [
        {
          function: 'filterFactory.logic.and',
          args: [
            {
              function: 'filterFactory.members',
              args: ['DM.Country.Country', ['United States']],
            },
            {
              function: 'filterFactory.members',
              args: ['DM.Brand.Brand', ['Brand A']],
            },
          ],
        },
      ];

      expect(() => {
        translateMeasuresFromJSONFunctionCall(
          mockMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow('Invalid measure JSON');
    });
  });

  describe('translateMeasuresFromJSON', () => {
    it('should return empty array when measuresJSON is null', () => {
      const measures = translateMeasuresJSON(
        null as any,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(measures).toEqual([]);
    });

    it('should return empty array when measuresJSON is undefined', () => {
      const measures = translateMeasuresJSON(
        undefined as any,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(measures).toEqual([]);
    });

    it('should return empty array when measuresJSON is false', () => {
      const measures = translateMeasuresJSON(
        false as any,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(measures).toEqual([]);
    });

    it('should return empty array when measuresJSON is 0', () => {
      const measures = translateMeasuresJSON(
        0 as any,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(measures).toEqual([]);
    });

    it('should return empty array when measuresJSON is empty string', () => {
      const measures = translateMeasuresJSON(
        '' as any,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(measures).toEqual([]);
    });

    it('should translate empty array to empty array', () => {
      const measures = translateMeasuresJSON(
        [],
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );
      expect(measures).toEqual([]);
    });

    it('should throw error for array of strings instead of function calls', () => {
      const invalidMeasuresJSON = ['DM.Commerce.Revenue', 'DM.Commerce.Cost'] as any;

      expect(() => {
        translateMeasuresJSON(
          invalidMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow(
        'Invalid measures JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should throw error for array of objects missing function property', () => {
      const invalidMeasuresJSON = [{ args: ['DM.Commerce.Revenue', 'Total Revenue'] }] as any;

      expect(() => {
        translateMeasuresJSON(
          invalidMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow(
        'Invalid measures JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should throw error for array of objects missing args property', () => {
      const invalidMeasuresJSON = [{ function: 'measureFactory.sum' }] as any;

      expect(() => {
        translateMeasuresJSON(
          invalidMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow(
        'Invalid measures JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should throw error for array containing non-objects', () => {
      const invalidMeasuresJSON = [
        { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue', 'Total Revenue'] },
        'not an object',
      ] as any;

      expect(() => {
        translateMeasuresJSON(
          invalidMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow(
        'Invalid measures JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should throw error for array containing null values', () => {
      const invalidMeasuresJSON = [
        { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue', 'Total Revenue'] },
        null,
      ] as any;

      expect(() => {
        translateMeasuresJSON(
          invalidMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow(
        'Invalid measures JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should throw error for array containing boolean values', () => {
      const invalidMeasuresJSON = [
        { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue', 'Total Revenue'] },
        true,
      ] as any;

      expect(() => {
        translateMeasuresJSON(
          invalidMeasuresJSON,
          MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
          MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
        );
      }).toThrow(
        'Invalid measures JSON. Expected an array of function calls with "function" and "args" properties.',
      );
    });

    it('should translate valid parsed function call array', () => {
      const validMeasuresJSON = [
        { function: 'measureFactory.sum', args: ['DM.Commerce.Revenue', 'Total Revenue'] },
        { function: 'measureFactory.count', args: ['DM.Commerce.Revenue'] },
      ];

      const measures = translateMeasuresJSON(
        validMeasuresJSON,
        MOCK_DATA_SOURCE_SAMPLE_ECOMMERCE,
        MOCK_NORMALIZED_TABLES_SAMPLE_ECOMMERCE,
      );

      expect(Array.isArray(measures)).toBe(true);
      expect(measures).toHaveLength(2);
    });
  });
});
