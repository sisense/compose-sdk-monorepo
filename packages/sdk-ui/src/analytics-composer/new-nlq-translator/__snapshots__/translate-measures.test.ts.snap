// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translateMeasures > translateMeasuresFromJSONFunctionCall > should translate measures from parsed function calls 1`] = `
[
  {
    "__serializable": "DimensionalBaseMeasure",
    "aggregation": "sum",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Commerce.Revenue",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Commerce.Revenue]",
      "name": "Revenue",
      "type": "numeric-attribute",
    },
    "composeCode": "measureFactory.sum(DM.Commerce.Revenue, 'Total Revenue')",
    "dataSource": undefined,
    "description": "",
    "name": "Total Revenue",
    "sort": 0,
    "type": "basemeasure",
  },
  {
    "__serializable": "DimensionalBaseMeasure",
    "aggregation": "count",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Commerce.Revenue",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Commerce.Revenue]",
      "name": "Revenue",
      "type": "numeric-attribute",
    },
    "composeCode": "measureFactory.count(DM.Commerce.Revenue, 'Revenue Count')",
    "dataSource": undefined,
    "description": "",
    "name": "Revenue Count",
    "sort": 0,
    "type": "basemeasure",
  },
]
`;

exports[`translateMeasures > translateMeasuresFromJSONFunctionCall > should translate single measure 1`] = `
[
  {
    "__serializable": "DimensionalBaseMeasure",
    "aggregation": "avg",
    "attribute": {
      "__serializable": "DimensionalAttribute",
      "composeCode": "DM.Commerce.Revenue",
      "dataSource": {
        "address": "LocalHost",
        "id": "localhost_aSampleIAAaECommerce",
        "live": false,
        "title": "Sample ECommerce",
      },
      "description": "",
      "expression": "[Commerce.Revenue]",
      "name": "Revenue",
      "type": "numeric-attribute",
    },
    "composeCode": "measureFactory.average(DM.Commerce.Revenue)",
    "dataSource": undefined,
    "description": "",
    "name": "avg Revenue",
    "sort": 0,
    "type": "basemeasure",
  },
]
`;
